ext/rotoscope/rotoscope.c

```diff
@@ -144,24 +147,69 @@ static bool tracecmp(rs_tracepoint_t *a, rs_tracepoint_t *b) {
           a->method_level == b->method_level);
 }
 
-static void log_raw_trace(FILE *stream, rs_tracepoint_t trace) {
-  fprintf(stream, RS_CSV_FORMAT "\n", RS_CSV_VALUES(trace));
+static bool endof_block_method(rb_trace_arg_t *trace_arg) {
+  VALUE method = rb_obj_method(rb_tracearg_self(trace_arg),
+                               rb_tracearg_method_id(trace_arg));
+  VALUE iseq = rb_funcall(cInstructionSeq, id_of, 1, method);
+  if (!RTEST(iseq)) return false;
+
+  VALUE label = rb_funcall(iseq, id_label, 0);
+  char *label_str = StringValueCStr(label);
+  return strncmp(VM_BLOCK_PREFIX, label_str, strlen(VM_BLOCK_PREFIX)) == 0;
+}
+
+static void log_trace_event_with_caller(char *buffer, Rotoscope *config,
+                                        rs_stack_frame_t frame,
+                                        rb_event_flag_t event) {
+  if (METHOD_CALL_P(event)) {
+    sprintf(buffer, RS_FLATTENED_CSV_FORMAT, RS_FLATTENED_CSV_VALUES(frame));
+    if (!rs_strmemo_uniq(&config->call_memo, (unsigned char *)buffer)) {
+      *buffer = '\0';
+    }
+  }
 }
 
-unsigned char output_buffer[500];
-static void log_stack_frame(FILE *stream, rs_stack_t *stack,
-                            rs_strmemo_t **call_memo, rs_tracepoint_t trace,
-                            rb_event_flag_t event) {
-  if (event & EVENT_CALL) {
-    rs_stack_frame_t frame = rs_stack_push(stack, trace);
-    sprintf((char *)output_buffer, RS_FLATTENED_CSV_FORMAT "\n",
-            RS_FLATTENED_CSV_VALUES(frame));
+static void log_trace_event(char *buffer, Rotoscope *config,
+                            rs_tracepoint_t trace, rb_event_flag_t event) {
+  if (METHOD_CALL_P(event)) {
+    sprintf(buffer, RS_CSV_FORMAT, RS_CSV_VALUES(trace));
```

Comment:
`sprintf` will overrun the buffer if we write more than 500 chars. You could use `snprintf` to specify the max length of the buffer which will be safer (note that the length specified doesn't include the null terminator, so in this case the max length would be 499).